### PR TITLE
Smooth scrolling support (+mouse wheel scrolling in some terminals)

### PR DESCRIPTION
**Pull request by H3rnand3zzz**
Added mouse scroll support with shift+scroll using readline.

Points before release:

- [x] Mouse compatibility
- [x] @sjaeckel's suggestions
- [x] Infinite scroll support

Known Issues: Sometimes it gets incredibly laggy (probably have something to do with profanity core)

### CODE REVIEW

File: src/ui/inputwin.c [hunk 1]
```diff
@@ -391,7 +392,7 @@ _inp_edited(const wint_t ch)
     bytes[utf_len] = '\0';
     gunichar unichar = g_utf8_get_char(bytes);
 
-    return g_unichar_isprint(unichar);
+    return g_unichar_isprint(unichar) && (ch != KEY_MOUSE);
```
Comment (H3rnand3zzz) [inline src/ui/inputwin.c:395]: should be removed

File: src/ui/inputwin.c [hunk 2]
```diff
@@ -924,6 +927,45 @@ _inp_rl_subwin_pagedown_handler(int count, int key)
     return 0;
 }
 
+static int
+_inp_rl_scroll(int count, int key)
+{
+    if (!prefs_get_boolean(PREF_MOUSE))
+        return 0;
+    cons_show(g_strdup_printf("scroll. count: %d, key: %d, %c", count, key, key));
```
Comment (sjaeckel) [inline src/ui/inputwin.c:935]: I guess this in only a debug statement, which will go soonish.

FYI `cons_show()` behaves like `printf()`, i.e. you can drop the call to `g_strdup_printf()`. (also you're leaking the value returned :D)

**Comment by H3rnand3zzz** [comment]
It's ready for review, but not ready for release. I would appreciate advises on lag issue. Probably core-related. Even without binding, mouse scrolling makes considerable lag.

File: src/ui/inputwin.c [hunk 3]
```diff
@@ -927,6 +935,44 @@ _inp_rl_subwin_pagedown_handler(int count, int key)
...
+}
+
+static int
+_inp_rl_scroll_impl(int count, int key)
+{
+    if (!prefs_get_boolean(PREF_MOUSE))
+        return 0;
+    ProfWin* window = wins_get_current();
+
+    if (key == 'B') { // mouse wheel down
```
Comment (jubalh) [inline src/ui/inputwin.c:968]: here i agree with clang that the comments should be on top or inside ;)

File: src/command/cmd_defs.c [hunk 4]
```diff
@@ -1283,6 +1283,22 @@ static const struct cmd_t command_defs[] = {
...
+      CMD_SYN(
+              "/mouse on",
+              "/mouse off")
+      CMD_DESC(
+              "Mouse scroll controls.")
+      CMD_ARGS(
+              { "on", "Activate mouse scroll." },
+              { "off", "Deactivate mouse scroll." })
+    },
+    //
```
Comment (jubalh) [inline src/command/cmd_defs.c:1300]: ?

**Review by jubalh — CHANGES_REQUESTED** [review]

File: src/command/cmd_defs.c (hunk 4 above)
Comment (sjaeckel) [inline src/command/cmd_defs.c:1300]:
```suggestion
```

File: src/ui/inputwin.c [hunk 5]
```diff
@@ -927,6 +935,46 @@ _inp_rl_subwin_pagedown_handler(int count, int key)
...
+_inp_rl_scroll_trampoline(int count, int key)
+{
+    return _inp_rl_scroll_fp(count, key);
+}
+
+static int
+_inp_rl_scroll_impl(int count, int key)
+{
+    if (!prefs_get_boolean(PREF_MOUSE))
+        return 0;
```
Comment (sjaeckel) [inline src/ui/inputwin.c:965]:
```suggestion
```
This can go, since this function will only be called if this flag is set to true.

**Review by sjaeckel — CHANGES_REQUESTED** [review]
Again clean and easy to follow.

I'd like to have a second look tomorrow after those proposed changes are included.

**Comment by H3rnand3zzz** [comment]
To make it clear

At the current stage, the commit is incompatible with mouse and works only with touch pad. I hope to resolve it as soon as possible.

**Comment by H3rnand3zzz** [comment]
Rebased on master
Followed by changing for alt+up/down instead of shift+up/down. Alt it more intuitive (since it presumes ALTernative behaviour for the same button) and minor changes to merge with previous updates.

**Review by jubalh — APPROVED** [review]

**Comment by H3rnand3zzz** [comment]
@sjaeckel, is PR ok now?

**Comment by sjaeckel** [comment]
I'll have a look soon, currently only on mobile

**Comment by H3rnand3zzz** [comment]
### New revision
 - Changes requested by @sjaeckel are implemented:
   - Removed preference (since alt+arrows support for navigation does not need to be optional)
   - Details are added in command description regarding other terminals' support

**Review by sjaeckel — APPROVED** [review]

**Pull request merged**